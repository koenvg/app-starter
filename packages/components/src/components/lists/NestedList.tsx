import React, { SFC } from "react";
import { ListItem, Collapse, List } from "@material-ui/core";
import ExpandLess from "@material-ui/icons/ExpandLess";
import ExpandMore from "@material-ui/icons/ExpandMore";

export interface Props {
  header: JSX.Element;
  className?: string;
  dense?: boolean;
  startOpen?: boolean;
}

export const NestedList: SFC<Props> = ({ children, header, className, dense, startOpen = false }) => {
  const [open, setIsOpen] = React.useState(startOpen);
  const handleClick = React.useCallback(() => {
    setIsOpen(value => !value);
  }, []);

  return (
    <>
      <ListItem button onClick={handleClick} className={className}>
        {header}
        {open ? <ExpandLess /> : <ExpandMore />}
      </ListItem>
      <Collapse in={open} timeout="auto" unmountOnExit>
        <List component="div" disablePadding dense={dense}>
          {children}
        </List>
      </Collapse>
    </>
  );
};

NestedList.displayName = "NestedList";
