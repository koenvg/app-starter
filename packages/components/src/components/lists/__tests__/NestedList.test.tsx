import { NestedList, Props } from "../NestedList";
import React from "react";
import { render, fireEvent, waitForElementToBeRemoved } from "@testing-library/react";
import { ListItemText, ListItem } from "@material-ui/core";

describe("NestedList", () => {
  const defaultProps: Props = {
    header: <ListItemText>Header</ListItemText>,
  };
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("render the nested list closed by default", () => {
    const { container } = render(
      <NestedList {...defaultProps}>
        <ListItem>
          <ListItemText>Nested</ListItemText>
        </ListItem>
      </NestedList>,
    );
    expect(container).toMatchSnapshot();
  });

  it("should open and close the nested list when clicking the header item", async () => {
    const { container, queryByText, getByText } = render(
      <NestedList {...defaultProps}>
        <ListItem>
          <ListItemText>Nested</ListItemText>
        </ListItem>
      </NestedList>,
    );

    fireEvent.click(getByText("Header"));

    expect(container).toMatchSnapshot();
    fireEvent.click(getByText("Header"));

    await waitForElementToBeRemoved(() => queryByText("Nested"));
  });
});
