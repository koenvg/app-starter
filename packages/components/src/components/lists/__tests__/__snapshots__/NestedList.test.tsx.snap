// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedList render the nested list closed by default 1`] = `
<div>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
      >
        Header
      </span>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
</div>
`;

exports[`NestedList should open and close the nested list when clicking the header item 1`] = `
<div>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
      >
        Header
      </span>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
  <div
    class="MuiCollapse-container"
    style="min-height: 0px; height: 0px; transition-duration: 0ms;"
  >
    <div
      class="MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner"
      >
        <div
          class="MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                Nested
              </span>
            </div>
          </li>
        </div>
      </div>
    </div>
  </div>
</div>
`;
